src/Razor/src/Microsoft.CodeAnalysis.Remote.Razor/ProjectSystem/ProjectQueryServiceFactory.cs

```diff
@@ -0,0 +1,87 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Composition;
+using Microsoft.AspNetCore.Razor.PooledObjects;
+using Microsoft.CodeAnalysis.Razor.ProjectSystem;
+
+namespace Microsoft.CodeAnalysis.Remote.Razor.ProjectSystem;
+
+[Export(typeof(ProjectQueryServiceFactory)), Shared]
+[method: ImportingConstructor]
+internal sealed class ProjectQueryServiceFactory(ProjectSnapshotFactory projectSnapshotFactory)
+{
+    public IProjectQueryService Create(Solution solution)
+        => new Service(solution, projectSnapshotFactory);
+
+    private sealed class Service(Solution solution, ProjectSnapshotFactory projectSnapshotFactory) : IProjectQueryService
+    {
+        private readonly Solution _solution = solution;
+        private readonly ProjectSnapshotFactory _projectSnapshotFactory = projectSnapshotFactory;
+
+        private ImmutableArray<IProjectSnapshot> _projects;
+
+        public ImmutableArray<IProjectSnapshot> GetProjects()
+        {
+            if (_projects.IsDefault)
+            {
+                ImmutableInterlocked.InterlockedInitialize(ref _projects, ComputeProjects(_solution, _projectSnapshotFactory));
+            }
+
+            return _projects;
+
+            static ImmutableArray<IProjectSnapshot> ComputeProjects(Solution solution, ProjectSnapshotFactory factory)
+            {
+                var projectIds = solution.ProjectIds;
+
+                if (projectIds.Count == 0)
+                {
+                    return [];
+                }
+
+                using var results = new PooledArrayBuilder<IProjectSnapshot>(capacity: projectIds.Count);
+
+                foreach (var projectId in projectIds)
+                {
+                    if (solution.GetProject(projectId) is Project project)
+                    {
+                        results.Add(factory.GetOrCreate(project));
```

Comment:
Whilst it would still be possible for this to create 100s of project snapshots for a large solution, it feels a little wasteful to pro-actively do it. In future with cohosting the CWT that these are stored in should go away, so they will be stateless wrappers around `Project` (hopefully) so maybe it won't feel as bad then, but right now changing this from `IEnumerable` to `ImmutableArray` gives me pause.

Maybe this is just an issue to log for future work, to remove this method entirely and only allow methods like the below, which can do some filtering, to limit their scope.

Reply:
That's a fair point. I don't like `IEnumerable` all that much here but it might be the best choice. I'll take another look.

Reply:
Maybe naming the method `GetRazorProjects()` and filtering to only projects that contain Razor files would be a reasonable compromise?

Reply:
It might. Wouldn't that theoretically match what's be found in the `IProjectSnapshotManager`?

Reply:
I ended up switching this back to `IEnumerable` and made co-hosting enforce that `RemoteDocumentSnapshot` and `RemoteProjectSnapshot` are only ever created for Razor documents and projects containing Razor documents, respectively.